### PR TITLE
RBAC CLI: role subcommands (create, update, delete, and show)

### PR DESCRIPTION
**Pull request by peterschwarz**
This PR adds the `role` sub-commands `show`, `create`, `update`, `delete` to the CLI.

# Testing

## Setup a local splinter node

These operations for setup can be done in one terminal.

First, initialize your splinter instance:

```
$ SPLINTER_HOME=/tmp/splinter cargo run --manifest-path cli/Cargo.toml --features experimental -- cert generate
$ SPLINTER_HOME=/tmp/splinter cargo run --manifest-path cli/Cargo.toml --features experimental -- database migrate
```

Also, generate a key and add it to the allow_keys file:
```
$ cargo run --manifest-path cli/Cargo.toml --features experimental -- keygen
$ cat ~/.cylinder/keys/$USER.pub >> /tmp/splinter/data/allow_keys
```

Start your splinter node:
```
SPLINTER_HOME=/tmp/splinter cargo run --manifest-path splinterd/Cargo.toml --features experimental -- --tls-insecure --enable-biome -vv --rest-api-endpoint http://127.0.0.1:8080
```

## Create a Role

These remainder of the operations in these test instructions can be done in a second terminal. 

List the current roles:

```
$ cargo run --manifest-path cli/Cargo.toml --features experimental -- \
   role list
ID  DISPLAY NAME
```
(if you are testing on node with existing roles, this may look different).

Add a role:
```
$  cargo run --manifest-path cli/Cargo.toml --features experimental -- \
    role create \
    --display "Status Reader" \
    --permission "status.read" \
     test_role
```

Now, list the roles again:

```
$ cargo run --manifest-path cli/Cargo.toml --features experimental -- \
   role list
ID        DISPLAY NAME
test_role Status Reader
```

## Show a Role

Display the role just created:

```
$ cargo run --manifest-path cli/Cargo.toml --features experimental -- \
   role show test_role
Id: test_role
    Display Name: Status Reader
    Permissions:
        status.read
```
## Update the role

Let's change the role to be about circuit reads:
```
$ cargo run --manifest-path cli/Cargo.toml --features experimental -- \
   role update \
   --rm-perm status.read \
   --add-perm circuit-read \
   --display "Circuit Reader" \
   test_role
```
And display it again:
```
$ cargo run --manifest-path cli/Cargo.toml --features experimental -- role show test_role
Id: test_role
    Display Name: Circuit Reader
    Permissions:
        circuit-read
```

## Delete the role

```
$ cargo run --manifest-path cli/Cargo.toml --features experimental -- \
   role delete test_role
```

Verify the deletion:
```
$ cargo run --manifest-path cli/Cargo.toml --features experimental -- \
   role show test_role
ERROR: Subcommand encountered an error: Failed to get role test_role: Role not found
```

### CODE REVIEW

File: cli/src/action/api/rbac.rs [hunk 1]
```diff
@@ -413,6 +413,37 @@ pub fn update_role(base_url: &str, auth: &str, role_update: RoleUpdate) -> Resul
...
+        .and_then(|res| {
+            let status = res.status();
+            if status.is_success() {
+                Ok(())
+            } else {
+                let message = res
+                    .json::<super::ServerError>()
+                    .map_err(|_| {
+                        CliError::ActionError(format!(
+                            "Get role delete request failed with status code '{}', but error \
```
Comment (shannynalayna) [inline cli/src/action/api/rbac.rs:432]: This error message looks a little wonky - seems like a `GET` request was just made

File: cli/src/action/api/rbac.rs [hunk 2]
```diff
@@ -43,6 +43,68 @@ impl fmt::Display for Role {
     }
 }
 
+/// Constructs roles for submission to a splinter node.
```
Comment (shannynalayna) [inline cli/src/action/api/rbac.rs:46]: Typo in commit message 'changes' -> 'change'

File: cli/man/splinter-role-create.1.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+<!--
+  Copyright 2018-2021 Cargill Incorporated
+  Licensed under Creative Commons Attribution 4.0 International License
+  https://creativecommons.org/licenses/by/4.0/
+-->
+
+NAME
+====
+
+**splinter-role-create** — Displays information about a role
```
Comment (agunde406) [inline cli/man/splinter-role-create.1.md:11]: Copy pasta

File: cli/man/splinter-role-update.1.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+<!--
+  Copyright 2018-2021 Cargill Incorporated
+  Licensed under Creative Commons Attribution 4.0 International License
+  https://creativecommons.org/licenses/by/4.0/
+-->
+
+NAME
+====
+
+**splinter-role-update** — Displays information about a role
```
Comment (agunde406) [inline cli/man/splinter-role-update.1.md:11]: copypasta

File: cli/man/splinter-role-update.1.md [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+This example updates the `circuit_admin` role on a splinter node.
+
+* The role has ID `circuit_admin`.
+* The role has a display name of `"Circuit Admin"`
+* The role has the permissions circuit read and write
+* The role will be updated to have the permission `status.read`
+
+
+```
+$ splinter role create \
```
Comment (agunde406) [inline cli/man/splinter-role-update.1.md:75]: Should say `update` not `create`

**Review by agunde406 — CHANGES_REQUESTED** [review]
Tested

File: cli/man/splinter-role-create.1.md (hunk 3 above)
Comment (agunde406) [inline cli/man/splinter-role-create.1.md:11]: copy pasta

**Review by agunde406 — CHANGES_REQUESTED** [review]

File: cli/src/action/api/rbac.rs (hunk 2 above)
Comment (peterschwarz) [reply cli/src/action/api/rbac.rs:46]: Updated.

File: cli/src/action/api/rbac.rs (hunk 1 above)
Comment (peterschwarz) [reply cli/src/action/api/rbac.rs]: Fixed

File: cli/man/splinter-role-update.1.md (hunk 4 above)
Comment (peterschwarz) [reply cli/man/splinter-role-update.1.md]: fixed

File: cli/man/splinter-role-update.1.md (hunk 5 above)
Comment (peterschwarz) [reply cli/man/splinter-role-update.1.md]: fixed

File: cli/man/splinter-role-create.1.md (hunk 3 above)
Comment (peterschwarz) [reply cli/man/splinter-role-create.1.md]: fixed

**Review by agunde406 — APPROVED** [review]

**Comment by vaporos** [comment]
Suggest that in show, we change:

```
Id: test_role
    Display Name: Status Reader
    Permissions:
        status.read
```

to

```
test_role - Status Reader
    Permissions:
        status.read
```

or

```
test_role
    Name: Status Reader
    Permissions:
        status.read
```

**Comment by vaporos** [comment]
Currently, the command does not produce an error if the we attempt to remove a permission that doesn't exist, or if we attempt to add a permission that already exists. If this is intentional, this I suggest a) the command should print what it is updating and what didn't need to change; and b) we should have an option like --strict which prevents the command from running if one of these conditions occur. I think though, in that case, it might be better to flip it around and implement --force as the option which makes it succeed even if its asked to do things that don't apply.

**Comment by vaporos** [comment]
With the current command, it is not obvious how one would set specific permissions to a role that may already have permissions. This could be accomplished by adding something like "--rm-all" or "--clear" which removes all the ones that currently exists and then adds all the specified ones, so that effectively we get only the ones we are listing.

**Comment by vaporos** [comment]
ERROR: Subcommand encountered an error: Failed to get role test_role: Role not found

should really be simple, like:

ERROR: Role does not exist

**Review by vaporos — APPROVED** [review]

**Comment by peterschwarz** [comment]
Updated the error messages, will update the `update` command arguments in a followup PR.

**Comment by peterschwarz** [comment]
Also, added a commit to change "Display Name" to "Name" in the list output.  The show command output has been corrected similarly, but in their original commits

**Review by shannynalayna — APPROVED** [review]
Tested

**Review by agunde406 — APPROVED** [review]

**Pull request merged**